File: mkindayzir/src/services/report.service.ts
import prisma from "@/lib/prisma";
import { requirePermission, PERMISSIONS } from "@/lib/rbac";

function authError(result: { authorized: boolean; error?: any }) {
  if (!result.authorized && result.error) {
    throw result.error;
  }
  return result;
}

export class ReportService {
  async getDashboardSummary(userId: string, user: { id: string; role: string }) {
    const auth = await requirePermission(PERMISSIONS.VIEW_DASHBOARD);
    authError(auth);

    try {
      const [totalProjects, openWorkItems, assignedToMe, overdueItems] = await Promise.all([
        prisma.project.count({ where: { status: "ACTIVE" } }),
        prisma.workItem.count({ where: { status: { not: "done" } } }),
        prisma.workItem.count({ where: { assigneeId: userId, status: { not: "done" } } }),
        prisma.workItem.count({
          where: {
            status: { not: "done" },
            dueDate: { lt: new Date() },
          },
        }),
      ]);

      return { totalProjects, openWorkItems, assignedToMe, overdueItems };
    } catch (error) {
      console.error("ReportService.getDashboardSummary error:", error);
      throw error;
    }
  }

  async getWorkloadReport(user: { id: string; role: string }) {
    const auth = await requirePermission(PERMISSIONS.VIEW_DASHBOARD);
    authError(auth);

    try {
      const workItems = await prisma.workItem.findMany({
        where: { status: { not: "done" }, assigneeId: { not: null } },
        include: {
          assignee: { select: { id: true, displayName: true, email: true } },
          project: { select: { id: true, key: true, name: true } },
        },
      });

      const grouped = workItems.reduce<Record<string, { assignee: { id: string; displayName: string; email: string }; items: typeof workItems }>>((acc, item) => {
        const assigneeId = item.assigneeId!;
        if (!acc[assigneeId]) {
          acc[assigneeId] = {
            assignee: {
              id: item.assignee!.id,
              displayName: item.assignee!.displayName,
              email: item.assignee!.email,
            },
            items: [],
          };
        }
        acc[assigneeId].items.push(item);
        return acc;
      }, {});

      return Object.values(grouped).map((group) => ({
        ...group,
        count: group.items.length,
      }));
    } catch (error) {
      console.error("ReportService.getWorkloadReport error:", error);
      throw error;
    }
  }

  async getVelocityReport(user: { id: string; role: string }, projectId?: string) {
    const auth = await requirePermission(PERMISSIONS.VIEW_DASHBOARD);
    authError(auth);

    try {
      const where: Record<string, unknown> = {
        status: "done",
        iterationId: { not: null },
      };

      if (projectId) {
        where.projectId = projectId;
      }

      const workItems = await prisma.workItem.findMany({
        where,
        include: {
          iteration: { select: { id: true, name: true, startDate: true, endDate: true } },
          project: { select: { id: true, key: true, name: true } },
        },
      });

      const grouped = workItems.reduce<Record<string, { iteration: { id: string; name: string; startDate: string; endDate: string }; totalPoints: number; count: number }>>((acc, item) => {
        if (!item.iterationId) return acc;
        if (!acc[item.iterationId]) {
          acc[item.iterationId] = {
            iteration: {
              id: item.iteration!.id,
              name: item.iteration!.name,
              startDate: item.iteration!.startDate!.toISOString(),
              endDate: item.iteration!.endDate!.toISOString(),
            },
            totalPoints: 0,
            count: 0,
          };
        }
        acc[item.iterationId].totalPoints += item.storyPoints ?? 0;
        acc[item.iterationId].count += 1;
        return acc;
      }, {});

      return Object.values(grouped);
    } catch (error) {
      console.error("ReportService.getVelocityReport error:", error);
      throw error;
    }
  }

  async getTrendReport(user: { id: string; role: string }) {
    const auth = await requirePermission(PERMISSIONS.VIEW_DASHBOARD);
    authError(auth);

    try {
      const thirtyDaysAgo = new Date();
      thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);

      const [createdItems, resolvedItems] = await Promise.all([
        prisma.workItem.findMany({
          where: {
            createdAt: { gte: thirtyDaysAgo },
          },
          select: {
            createdAt: true,
          },
        }),
        prisma.workItem.findMany({
          where: {
            status: "done",
            resolvedAt: { gte: thirtyDaysAgo },
          },
          select: {
            resolvedAt: true,
          },
        }),
      ]);

      const daily: Record<string, { date: string; created: number; resolved: number }> = {};

      for (let i = 0; i < 30; i++) {
        const d = new Date();
        d.setDate(d.getDate() - i);
        const key = d.toISOString().split("T")[0];
        daily[key] = { date: key, created: 0, resolved: 0 };
      }

      createdItems.forEach((item) => {
        const key = item.createdAt.toISOString().split("T")[0];
        if (daily[key]) {
          daily[key].created += 1;
        }
      });

      resolvedItems.forEach((item) => {
        if (item.resolvedAt) {
          const key = item.resolvedAt.toISOString().split("T")[0];
          if (daily[key]) {
            daily[key].resolved += 1;
          }
        }
      });

      return Object.values(daily).sort((a, b) => a.date.localeCompare(b.date));
    } catch (error) {
      console.error("ReportService.getTrendReport error:", error);
      throw error;
    }
  }

  async exportCSV(user: { id: string; role: string }, filters?: { projectId?: string; status?: string }) {
    const auth = await requirePermission(PERMISSIONS.VIEW_DASHBOARD);
    authError(auth);

    try {
      const where: Record<string, unknown> = {};
      if (filters?.projectId) where.projectId = filters.projectId;
      if (filters?.status) where.status = filters.status;

      const workItems = await prisma.workItem.findMany({
        where,
        include: {
          project: { select: { key: true, name: true } },
          assignee: { select: { displayName: true, email: true } },
        },
        orderBy: { createdAt: "desc" },
      });

      const headers = ["ID", "Title", "Type", "Status", "Priority", "Assignee", "Project", "Created At"];
      const rows = workItems.map((item) => [
        item.id,
        `"${item.title.replace(/"/g, '""')}"`,
        item.type,
        item.status,
        item.priority,
        `"${(item.assignee?.displayName || "").replace(/"/g, '""')}"`,
        `"${item.project?.key || ""}"`,
        item.createdAt.toISOString(),
      ]);

      const csv = "\uFEFF" + [headers.join(","), ...rows.map((r) => r.join(","))].join("\n");
      return csv;
    } catch (error) {
      console.error("ReportService.exportCSV error:", error);
      throw error;
    }
  }
}

File: backend/app/services/report_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from app.models.project import Project
from app.models.work_item import WorkItem
from datetime import datetime, timezone, timedelta


class ReportService:
    @staticmethod
    async def get_dashboard_summary(db: AsyncSession, user_id: str, user: dict) -> dict:
        total_projects = await db.execute(select(func.count()).select_from(Project).where(Project.status == "ACTIVE", Project.deletedAt.is_(None)))
        open_work_items = await db.execute(select(func.count()).select_from(WorkItem).where(WorkItem.status != "done", WorkItem.deletedAt.is_(None)))
        assigned_to_me = await db.execute(select(func.count()).select_from(WorkItem).where(WorkItem.assigneeId == user_id, WorkItem.status != "done", WorkItem.deletedAt.is_(None)))
        overdue_items = await db.execute(select(func.count()).select_from(WorkItem).where(WorkItem.status != "done", WorkItem.dueDate < datetime.now(timezone.utc), WorkItem.deletedAt.is_(None)))
        return {
            "totalProjects": total_projects.scalar_one(),
            "openWorkItems": open_work_items.scalar_one(),
            "assignedToMe": assigned_to_me.scalar_one(),
            "overdueItems": overdue_items.scalar_one(),
        }

    @staticmethod
    async def get_workload_report(db: AsyncSession, user: dict) -> list[dict]:
        result = await db.execute(
            select(WorkItem).where(WorkItem.status != "done", WorkItem.assigneeId.isnot(None), WorkItem.deletedAt.is_(None))
        )
        items = result.scalars().all()
        grouped = {}
        for item in items:
            aid = item.assigneeId
            if aid not in grouped:
                grouped[aid] = {"assignee": {"id": aid}, "items": [], "count": 0}
            grouped[aid]["items"].append({
                "id": item.id,
                "title": item.title,
                "status": item.status,
                "priority": item.priority,
                "projectId": item.projectId,
            })
            grouped[aid]["count"] += 1
        return list(grouped.values())

    @staticmethod
    async def get_velocity_report(db: AsyncSession, user: dict, project_id: str | None = None) -> list[dict]:
        query = select(WorkItem).where(WorkItem.status == "done", WorkItem.iterationId.isnot(None), WorkItem.deletedAt.is_(None))
        if project_id:
            query = query.where(WorkItem.projectId == project_id)
        result = await db.execute(query)
        items = result.scalars().all()
        grouped = {}
        for item in items:
            if not item.iterationId:
                continue
            if item.iterationId not in grouped:
                grouped[item.iterationId] = {
                    "iteration": {"id": item.iterationId},
                    "totalPoints": 0,
                    "count": 0,
                }
            grouped[item.iterationId]["totalPoints"] += item.storyPoints or 0
            grouped[item.iterationId]["count"] += 1
        return list(grouped.values())

    @staticmethod
    async def get_trend_report(db: AsyncSession, user: dict) -> list[dict]:
        thirty_days_ago = datetime.now(timezone.utc) - timedelta(days=30)
        created = await db.execute(
            select(WorkItem.createdAt).where(WorkItem.createdAt >= thirty_days_ago, WorkItem.deletedAt.is_(None))
        )
        created_dates = [r[0].date().isoformat() for r in created.all()]
        resolved = await db.execute(
            select(WorkItem.resolvedAt).where(WorkItem.status == "done", WorkItem.resolvedAt >= thirty_days_ago, WorkItem.deletedAt.is_(None))
        )
        resolved_dates = [r[0].date().isoformat() for r in resolved.all() if r[0]]

        daily = {}
        for i in range(30):
            d = (datetime.now(timezone.utc) - timedelta(days=i)).date().isoformat()
            daily[d] = {"date": d, "created": 0, "resolved": 0}
        for d in created_dates:
            if d in daily:
                daily[d]["created"] += 1
        for d in resolved_dates:
            if d in daily:
                daily[d]["resolved"] += 1
        return sorted(daily.values(), key=lambda x: x["date"])

    @staticmethod
    async def export_csv(db: AsyncSession, user: dict, filters: dict | None = None) -> str:
        query = select(WorkItem)
        if filters:
            if filters.get("projectId"):
                query = query.where(WorkItem.projectId == filters["projectId"])
            if filters.get("status"):
                query = query.where(WorkItem.status == filters["status"])
        query = query.where(WorkItem.deletedAt.is_(None)).order_by(WorkItem.createdAt.desc())
        result = await db.execute(query)
        items = result.scalars().all()
        headers = ["ID", "Title", "Type", "Status", "Priority", "Assignee", "Project", "Created At"]
        rows = []
        for item in items:
            rows.append([item.id, item.title, item.type, item.status, item.priority, item.assigneeId or "", item.projectId, item.createdAt.isoformat() if item.createdAt else ""])
        csv_content = "\n".join([",".join(headers)] + [",".join(str(c) for c in r) for r in rows])
        return csv_content
